libr/anal/sign.c

```diff
@@ -774,6 +767,109 @@ R_API bool r_sign_add_anal(RAnal *a, const char *name, ut64 size, const ut8 *byt
 	return retval;
 }
 
+static RSignGraph *r_sign_fcn_graph(RAnalFunction *fcn) {
+	r_return_val_if_fail (fcn, false);
+	RSignGraph *graph = R_NEW0 (RSignGraph);
+	if (graph) {
+		graph->cc = r_anal_function_complexity (fcn),
+		graph->nbbs = r_list_length (fcn->bbs);
+		graph->edges = r_anal_function_count_edges (fcn, &graph->ebbs);
+		graph->bbsum = r_anal_function_realsize (fcn);
+	}
+	return graph;
+}
+
+static RSignBytes *r_sign_fcn_bytes(RAnal *a, RAnalFunction *fcn) {
+	r_return_val_if_fail (a && fcn, false);
+	RCore *core = a->coreb.core;
+	int maxsz = a->coreb.cfggeti (core, "zign.maxsz");
+	int fcnlen = r_anal_function_realsize (fcn);
+	int size = R_MIN (core->io->addrbytes * fcnlen, maxsz);
+
+	// hopefully not necissary, but just in case
+	if (size <= 0) {
+		eprintf ("error: Negative size for bytes signature\n");
+		return NULL;
+	}
+
+	// get bytes part of signature
+	ut8 *bytes = malloc (size);
+	if (!bytes) {
+		return NULL;
+	}
+	if (r_io_nread_at (core->io, fcn->addr, bytes, size) < 0) {
+		eprintf ("error: failed to read at 0x%08" PFMT64x "\n", fcn->addr);
+		free (bytes);
+		return NULL;
+	}
+
+	// get mask part of signature
+	ut8 *mask = r_anal_mask (a, size, bytes, fcn->addr);
+
+	// get bytes sig holder
+	RSignBytes *sig = R_NEW0 (RSignBytes);
+	if (!mask) {
```

Comment:
move this check one line above, othewise its leaking `sig`

Reply:
I meant that to be `sig`. If `sig` is NULL then `sig->size = size;` will segfault. Mask should probably be checked too...

To simplify cleanup and save some lines of code I created `bytes_sig_free`. I then use that function in a few other places. This is done in commit a959981f5.